zipline/pipeline/factors/technical.py

```diff
@@ -683,3 +679,115 @@ def compute(self, today, assets, out, highs, lows, closes):
             )),
             2
         )
+
+
+class MovingAverageConvergenceDivergence(_ExponentialWeightedFactor):
+    """
+    Moving Average Convergence/Divergence (MACD)
+    https://en.wikipedia.org/wiki/MACD
+
+    A technical indicator originally developed by Gerald Appel in the late
+    1970's. MACD shows the relationship between two moving averages and
+    reveals changes in the strength, direction, momentum, and duration of a
+    trend in a stock's price.
+
+    **Default Inputs:** :data:`zipline.pipeline.data.USEquityPricing.close`
+    **Default Window Length:** Window length is automatically calculated as the
+    sum of slow_period and signal_period.
+
+    Parameters
+    ----------
+    fast_period : int >= 0, <= window_length
+        The window length for the "fast" EWMA. Default is 12.
+    slow_period : int >= 0, <= window_length
+        The window length for the "slow" EWMA. Default is 26.
+    signal_period' : int >= 0, <= slow_period
+        The window length for the signal line. Default is 9.
+
+    Returns
+    -------
+    MACD: The difference between "fast" EWMA and "slow" EWMA.
+    signal: The EWMA of the MACD line using `signal_period` as span.
+    hist: Difference between MACD and signal. (Divergence series)
+    """
+    inputs = [USEquityPricing.close]
+    params = ('fast_period', 'slow_period', 'signal_period')
+    outputs = ('MACD', 'signal', 'hist')
+
+    def __new__(cls,
+                fast_period=12,
+                slow_period=26,
+                signal_period=9,
+                *args,
+                **kwargs):
+        return super(MovingAverageConvergenceDivergence, cls).__new__(
+            cls,
+            fast_period=fast_period,
+            slow_period=slow_period,
+            signal_period=signal_period,
+            window_length=slow_period + signal_period - 1,
+            *args, **kwargs
+        )
+
+    def calculate_ewma(self, data, length):
+        decay_rate = 1.0 - (2.0 / (1.0 + length))
+        return average(data,
+                       axis=1,
+                       weights=self.weights(length, decay_rate))
+
+    def calculate_macd(self, col):
+        slow_EWMA = self.calculate_ewma(
+            rolling_window(
+                col,
+                self.params['slow_period']
```

Comment:
We should pull out all the params values once instead of multiple times per value.